Documentation/design-docs/standalone-gc-loading.md

```diff
@@ -0,0 +1,222 @@
+and have realized these same benefits. The existence of an interface and an implementation loadable
+from shared libraries has enabled RyuJIT in particular to be used as the code generator for both the
+CoreRT compiler and crossgen, while still being flexible enough to be tested using tools that implement
+very non-standard execution engines such as [SuperPMI](https://github.com/dotnet/coreclr/blob/master/src/ToolBox/superpmi/readme.txt).
+
+The below loading protocol is inspired directly by the JIT loader and many aspects of the GC loader are identical
+to what the JIT does when loading dynamic shared libraries.
+
+## Loading a Standalone GC
+
+Given that it is possible to create a GC that resides in a dynamic shared library, it is important
+that the runtime have a protocol for locating and loading such GCs. The JIT is capable of being loaded
+in this manner and, because of this, a significant amount of prior art exists for loading components
+for shared libraries from the file system. This specification is based heavily on the ways that a
+standalone JIT can be loaded.
+
+Fundamentally, the algorithm for loading a standalone GC consists of these steps:
+
+0. Identify whether or not we should be using a standalone GC at all.
+1. Identify *where* we should be looking for GC shared libraries.
+2. Identify *what* we are looking for - the specific name of a dynamic shared library that
+   will be loaded.
+3. Load the dynamic shared library and ask it to identify itself (name and version).
+4. Check that the version numbers are compatible.
+5. If so, initialize the GC and continue on with EE startup. If not, reject the dynamic shared library
+   and raise an appropriate user-visible error.
+
+The algorithm for initializing the DAC against a target process using a standalone GC consists of these steps:
+
+1. Identify whether or not the target process is using a standalone GC at all. If not, no further
+   checks are necessary.
+2. If so, inspect the version number of the standalone GC in the target process and determine whether
+   or not the DAC is compatible with that version. If not, present a notification of some kind
+   that the debugging experience will be degraded.
+3. Continue onwards.
+
+Each one of these steps will be explained in detail below.
+
+### Identifying candidate shared libraries
+
+The question of whether or not the EE should attempt to locate and load a standalone GC
+is answered by the EE's configuration system (`EEConfig`). EEConfig has the ability to
+query configuration information from a variety of sources, such a registry keys or
+environment variables. Using this subsystem, users can specify a specific environment
+variable to indicate that they are interested in loading a standalone GC.
+
+There are three configuration keys that control the behavior of loading a standalone GC:
+
+1. `UseStandaloneGC`, (environment variable `COMPLUS_UseStandaloneGC`). If set and equal to "1", the EE
```

Comment:
Why do we need 3 things? Can we have just one? `COMPlus_GCName`?

I believe that it is how the JIT loading works - there is a single key `COMPlus_JitName` that you set, and it gets passed to LoadLibrary. Nothing fancy.

Reply:
Sure. I modeled it with this comment in mind: https://github.com/dotnet/coreclr/blob/2f37c02dfa3b649d4d7322724c95ab08cb12cb6f/src/vm/codeman.cpp#L1536-L1537. Keeping it simple sounds good to me.

Reply:
`g_CLRJITPath` is meant for loading the JIT from a nuget package: You reference a nuget package with a different JIT in your project and it gets pulled in. It required changes in number of places across the board to make it happen... .

Reply:
I see - thanks!